### ISSUE TITLE
Compilation is 4-6x slower for Python 3.12 vs 3.11

### ISSUE THREAD

**Issue by Aksem** [Nuitka/Nuitka#2964]
**Type:** Question/Potential regression

**Description:** libcst library includes a (generated) [source file with a lot of dataclasses](https://github.com/Instagram/LibCST/blob/main/libcst/matchers/__init__.py). Compilation of this `libcst.matchers` module takes a lot more time(4-6x, depends on OS. 20 min vs 1h 20 min+) for Python 3.12 vs Python 3.11. Is it known and expected behaviour? If so, do you have maybe some tips how it can be improved from my side?

I found in logs that compilation of .c file is the problem:

```
3.11:

Nuitka-Scons: Running /usr/bin/ccache /usr/bin/clang -o module.libcst.matchers.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -w -fvisibility=hidden -fvisibility-inlines-hidden -O3 -D_XOPEN_SOURCE -D__NUITKA_NO_ASSERT__ -D_NUITKA_TRACE -D_NUITKA_STANDALONE -D_NUITKA_CONSTANTS_FROM_INCBIN -D_NUITKA_FROZEN=151 -D_NUITKA_EXE -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.11/lib/python3.11/site-packages/nuitka/build/inline_copy/zlib -I/opt/hostedtoolcache/Python/3.11.9/x64/include/python3.11 -I. -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.11/lib/python3.11/site-packages/nuitka/build/include -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.11/lib/python3.11/site-packages/nuitka/build/static_src -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.11/lib/python3.11/site-packages/nuitka/build/inline_copy/libbacktrace module.libcst.matchers.c took 613.82 seconds

3.12:

Nuitka-Scons: Running /usr/bin/ccache /usr/bin/clang -o module.libcst.matchers.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -w -fvisibility=hidden -fvisibility-inlines-hidden -O3 -D_XOPEN_SOURCE -D__NUITKA_NO_ASSERT__ -D_NUITKA_TRACE -D_NUITKA_STANDALONE -D_NUITKA_CONSTANTS_FROM_INCBIN -D_NUITKA_FROZEN=153 -D_NUITKA_EXE -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.12/lib/python3.12/site-packages/nuitka/build/inline_copy/zlib -I/opt/hostedtoolcache/Python/3.12.4/x64/include/python3.12 -I. -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.12/lib/python3.12/site-packages/nuitka/build/include -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.12/lib/python3.12/site-packages/nuitka/build/static_src -I~/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.12/lib/python3.12/site-packages/nuitka/build/inline_copy/libbacktrace module.libcst.matchers.c took 3892.37 seconds
```

**Environment:**
```
2.3.10
Commercial: None
Python: 3.12.4 (main, Jun 20 2024, 16:57:58) [GCC 11.4.0]
Flavor: GitHub Actions Python
Executable: /home/runner/.cache/pypoetry/virtualenvs/nuitka-bug-v58dnU5z-py3.12/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu (based on Debian) 22.04.4
Version C compiler: /usr/bin/gcc (gcc 11).
```
Pyenv, Poetry.
OS: windows-2022, ubuntu-22.04, macos-13
Python 3.12.4, 3.11.9

Nuitka version: 2.3.10, factory (both are tested in CI of repo with example)

**MRE:** https://github.com/Aksem/nuitka_bug (repository with configured CI)

**Packages:**

- `libcst==1.4.0`

**Build command:**
```
poetry run python -m nuitka \
            --clang \
            --standalone \
            --follow-imports \
            --assume-yes-for-downloads \
            --plugin-enable="pylint-warnings" \
            --warn-unusual-code \
            --warn-implicit-exceptions \
            --show-memory \
            --show-modules \
            --trace-execution \
            nuitka_bug/main.py
```
and `--mingw64` on windows

**Comment by kayhayen** [comment]
For exceptions, there is an open item for the code generation to not use functions which convert back and forth between the 3 items version (type, value, tb) and the value with it included. The later is 3.12 and the former is old Python. The code generation was not yet adapted to use new APIs that would abstract it and allow for leaner C code. Also performance at run time can be a lot worse if a lot of exceptions are dealt with, which e.g. driving an async loop can do.

For now, I recommend not using 3.12 for deployments. I expect it will be 2.5 or 2.6 that will have this TODO solved only, but I will leave this issue for tracking it.

**Comment by KRRT7** [comment]
solved for 2.4 already.
thanks @kayhayen !
